### PR TITLE
special page sorting options and inline grid placing

### PR DESCRIPTION
**Pull request by bublath**
I updated my old patches for the current dev branch - wondering if this could be considered useful now
1. sort-pages option
// Sort pages by "title", "author", "2page", "compact", "desc" 
// Comma separated list of the options above
// title  : sort pages alphabetically by title (t:}
// author : sort pages alphabetically by subtitle {st:} - if this is used together with title, only title is used
// 2page  : make sure songs with 2 pages are placed on even pages, so the full song is visible in a normal book without needing to turn a page.
//          a blankpage is added to align. Note: this option requires extra precessing time (whole songbook takes about 2x to generate)
// compact: modifier for 2page - instead of adding a blank page, another song is moved in front of this song to achieve even page alignment for 2-page songs
// desc   : modifier for title or author to achive a descending order (Z->A instead of A->Z)
2. Alow to set grid placing with directive in song (useful if just one song does not really fit with standard grid placing)

### CODE REVIEW

**Comment by sciurius** [comment]
I've cherrypicked the 'inline grid placing' but introduced a new directive `{diagrams}` to obsolete `{grid}`.
See https://www.chordpro.org/beta/directives-diagrams/ .

**Comment by sciurius** [comment]
I'm trying the following songs:
````
{title: B Song1 (one page)}
[A]Hello

{ns}
{title: A Song2 (one page)}
[A]Hello
[B]World

{ns}
{title: C Song3 (two pages)}
[A]Hello
{colb}
[B]World[C]
````
with config
````
{
    "pdf" : {
	"pagealign-songs" : 0,
	"sort-pages" : "2page",
	"csv" : {
            "fields" : [
		{ "name" : "title", "meta" : "title"      },
		{ "name" : "pages", "meta" : "pagerange"  },
            ],
	},
    },
    "contents" : [],
}
````
This gives the following order (according to the CSV output):
````
title;pages
B Song1 (one page);1
A Song2 (one page);2
C Song3 (two pages);3-4
````
I would have expected the two page song to start on an even page. What am I missing?

**Comment by bublath** [comment]
It's been a long time since I worked on this.
Looking at it again, it does not seem to work properly for all cases. The placement actually only works if the "compact" option is used and I have doubts that it properly works when the last song is a 2-page song and needs to be placed differently. 
Somewhere in the process I seem to have lost the functionality to add empty pages - maybe a mistake in the merging/rebasing, since I believe I had that going once.
In a nutshell: Needs some work
Let me know if you seriously consider to add this functionality, then I can for sure take a look at it.
I was also thinking if I should rather encapsulate most of the extra code into a function.

**Comment by sciurius** [comment]
One of the reasons I consider to add this functionality is because it is very well delimited. There is one piece of code (that, as you already suggested should be moved to a separate subroutine) that does a well-defined job: it shuffles the songs in the songbook. 
But there are some other small changes that you made that would need attention.

**Comment by bublath** [comment]
Closing as I created a new cleaned up PR
